server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -0,0 +1,176 @@
+import org.elasticsearch.transport.TransportResponseHandler;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+
+public class ClusterBootstrapService {
+
+    private static final Logger logger = LogManager.getLogger(ClusterBootstrapService.class);
+
+    // The number of master-eligible nodes which, if discovered, can be used to bootstrap the cluster. This setting is unsafe in the event
+    // that more master nodes are started than expected.
+    public static final Setting<Integer> INITIAL_MASTER_NODE_COUNT_SETTING =
+        Setting.intSetting("cluster.unsafe_initial_master_node_count", 0, 0, Property.NodeScope);
+
+    private final int initialMasterNodeCount;
+    private final TransportService transportService;
+    private volatile boolean running;
+
+    public ClusterBootstrapService(Settings settings, TransportService transportService) {
+        initialMasterNodeCount = INITIAL_MASTER_NODE_COUNT_SETTING.get(settings);
+        this.transportService = transportService;
+    }
+
+    public void start() {
+        assert running == false;
+        running = true;
+
+        if (initialMasterNodeCount > 0 && transportService.getLocalNode().isMasterNode()) {
+            logger.debug("unsafely waiting for discovery of [{}] master-eligible nodes", initialMasterNodeCount);
+
+            final ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
+            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                threadContext.markAsSystemContext();
+                awaitDiscovery();
+            }
+        }
+    }
+
+    public void stop() {
+        assert running == true;
+        running = false;
+    }
+
+    private void awaitDiscovery() {
+        if (running == false) {
+            logger.debug("awaitDiscovery: not running");
+            return;
+        }
+
+        final GetDiscoveredNodesRequest request = new GetDiscoveredNodesRequest();
+        request.setWaitForNodes(initialMasterNodeCount);
+        logger.trace("sending {}", request);
+        transportService.sendRequest(transportService.getLocalNode(), GetDiscoveredNodesAction.NAME, request,
+            new TransportResponseHandler<GetDiscoveredNodesResponse>() {
+                @Override
+                public void handleResponse(GetDiscoveredNodesResponse response) {
+                    assert response.getNodes().size() >= initialMasterNodeCount;
+                    assert response.getNodes().stream().allMatch(DiscoveryNode::isMasterNode);
+
+                    logger.debug("discovered {}, starting to bootstrap", response.getNodes());
+                    awaitBootstrap(response.getBootstrapConfiguration());
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    if (exp.getRootCause() instanceof ElasticsearchTimeoutException) {
```

Comment:
should we simply set the timeout to a very high value instead of adding this retry logic here?

Reply:
How high is high enough? I'd prefer to retry forever rather than have to remember that this might time out and stop in a future debugging session.

Reply:
We can also make the timeout optional, so that setting it to null makes it unbounded (i.e. does not schedule a timeout)

Reply:
Ok, introduced nullability to the timeout in 9f7e95161db8f035762124e5c0e44ef2e42bb209.